components/fxa-client/src/oauth.rs

```diff
@@ -93,11 +107,41 @@ impl FirefoxAccount {
     }
 
     /// Check whether user is authorized using our refresh token.
-    pub fn check_authorization_status(&self) -> Result<IntrospectInfo> {
+    pub fn check_authorization_status(&mut self) -> Result<IntrospectInfo> {
         let resp = match self.state.refresh_token {
-            Some(ref refresh_token) => self
-                .client
-                .oauth_introspect_refresh_token(&self.state.config, &refresh_token.token)?,
+            Some(ref refresh_token) => {
+                match self
+                    .client
+                    .oauth_introspect_refresh_token(&self.state.config, &refresh_token.token)
```

Comment:
In the specific bug we saw most recently in Fenix, this `oauth_introspect_refresh_token` call was actually returning `200 OK`, and it was the downstream call to the sync storage node that was giving the 401s, causing Fenix to try calling `check_authorization_status` over and over again.

Luckily, I think that makes things simpler for us here - we don't have to match on the response code, we just need to rate-limit *all* calls to this method if they occur within the configured time window. So I think you can just pull your circuit-breaker logic up to the top of the function, like this:

```
fn check_authorization_status(...) {
  // Increment `attempts` etc.
  // Return an error if we've exceeded the limit.
  // If below the limit, proceed with the body of the function as before.
}
```